### ISSUE TITLE
Logfile location for parsing port incorrect on MacOS

### ISSUE THREAD

**Issue by v1ctor** [spatialos/gdk-for-unity#1226]
#### Description
~FPS Starter Project can't connect to a deployment because~ it can't find a port in a logfile on MacOS.

> EDIT: Was a red herring, see below - https://github.com/spatialos/gdk-for-unity/issues/1226#issuecomment-558633346

I found in the code that the log file is being searched like this:

```csharp
case RuntimePlatform.OSXPlayer:
                    // On MacOS it always goes in the Unity folder regardless of what companyName and productName are set to.
                    logPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Library", "Logs", "Unity", "Player.log");
                    break;
```
Which is:
> /Users/username/Library/Logs/Unity/Player.log

However Launcher opens a completely different directory if you click on a "Open Log File" button:

> /Users/username/Library/Logs/**Improbable/FPS Starter Project**/Player.log"

## Expected behaviour

GDK should use the same directory for a log file as SpatialOS Launcher to find a correct port.

## Current behaviour
GDK uses wrong directory for a log file and couldn't find a port.

## Possible solution
I couldn't find a launcher code, is it open source?

## Steps to reproduce
Follow https://docs.improbable.io/unity/alpha/projects/fps/get-started/get-started on macOS Mojave

## Environment
I was trying to start with FPS Starter Project on macOS.
Unity 2019.2.6f1
macOS Mojave 10.14.6 (18G87

**Comment by jamiebrynes7** [comment]
Hey @v1ctor thanks for the report! 

As the comment in the source may suggest, this was abnormal when we originally wrote this feature. Its possible in the meantime Unity has fixed up the directory to align with the other platforms. I'll raise a bug internally to check it out! 😄

Did this prevent you from connecting to the deployment? 

As that is unexpected behaviour, this port is just used for debugging purposes not for connecting to the SpatialOS deployment. It should log the error, but that's all: 

```csharp
#if !UNITY_EDITOR && DEVELOPMENT_BUILD && !UNITY_ANDROID && !UNITY_IPHONE
            try
            {
                var port = GetPlayerConnectionPort();
                Worker.SendLogMessage(LogLevel.Info, $"Unity PlayerConnection port: {port}.", Worker.WorkerId, null);
            }
            catch (Exception e)
            {
                logger.HandleLog(LogType.Exception, new LogEvent("Could not find the Unity PlayerConnection port.").WithException(e));
            }
#endif
```

**Comment by v1ctor** [comment]
> Did this prevent you from connecting to the deployment?

Yes I couldn't connect to the deployment.

I'm getting this exception:
Exception: Could n

> ot find PlayerConnection port in logfile
>   at Improbable.Gdk.Core.WorkerConnector.ExtractPlayerConnectionPort (System.String fileContents) [0x0005d] in /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs:324 
>   at Improbable.Gdk.Core.WorkerConnector.GetPlayerConnectionPort () [0x00125] in /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs:300 
>   at Improbable.Gdk.Core.WorkerConnector+<Connect>d__9.MoveNext () [0x002f1] in /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs:138 
> UnityEngine.DebugLogHandler:Internal_LogException(Exception, Object)
> UnityEngine.DebugLogHandler:LogException(Exception, Object)
> UnityEngine.Logger:LogException(Exception, Object)
> Improbable.Gdk.Core.ForwardingDispatcher:HandleLog(LogType, LogEvent) (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Logging/ForwardingDispatcher.cs:76)
> Improbable.Gdk.Core.<Connect>d__9:MoveNext() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs:143)
> System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
> System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
> System.Threading.ExecutionContext:Run(ExecutionContext, ContextCallback, Object, Boolean)
> System.Runtime.CompilerServices.MoveNextRunner:Run()
> System.Threading.Tasks.AwaitTaskContinuation:InvokeAction(Object)
> System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
> System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation:Run(Task, Boolean)
> System.Threading.Tasks.Task:FinishContinuations()
> System.Threading.Tasks.Task:FinishStageThree()
> System.Threading.Tasks.Task`1:TrySetResult(WorkerInWorld)
> System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetResult(WorkerInWorld)
> Improbable.Gdk.Core.<ConnectWithRetries>d__12:MoveNext() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs:203)
> System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
> System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
> System.Threading.ExecutionContext:Run(ExecutionContext, ContextCallback, Object, Boolean)
> System.Runtime.CompilerServices.MoveNextRunner:Run()
> System.Threading.Tasks.AwaitTaskContinuation:InvokeAction(Object)
> System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
> System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation:Run(Task, Boolean)
> System.Threading.Tasks.Task:FinishContinuations()
> System.Threading.Tasks.Task:FinishStageThree()
> System.Threading.Tasks.Task`1:TrySetResult(WorkerInWorld)
> System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetResult(WorkerInWorld)
> Improbable.Gdk.Core.<CreateWorkerInWorldAsync>d__10:MoveNext() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerInWorld.cs:65)
> System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
> System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
> System.Threading.ExecutionContext:Run(ExecutionContext, ContextCallback, Object, Boolean)
> System.Runtime.CompilerServices.MoveNextRunner:Run()
> System.Threading.Tasks.AwaitTaskContinuation:InvokeAction(Object)
> System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
> System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation:Run(Task, Boolean)
> System.Threading.Tasks.Task:FinishContinuations()
> System.Threading.Tasks.Task:FinishStageThree()
> System.Threading.Tasks.Task`1:TrySetResult(IConnectionHandler)
> System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:SetResult(IConnectionHandler)
> Improbable.Gdk.Core.<CreateAsync>d__11:MoveNext() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/ConnectionHandlers/SpatialOSConnectionHandlerBuilder.cs:90)
> System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
> System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
> System.Threading.ExecutionContext:Run(ExecutionContext, ContextCallback, Object, Boolean)
> System.Runtime.CompilerServices.MoveNextRunner:Run()
> System.Threading.Tasks.<>c:<.cctor>b__7_0(Object)
> UnityEngine.WorkRequest:Invoke() (at /Users/builduser/buildslave/unity/build/Runtime/Export/Scripting/UnitySynchronizationContext.cs:114)
> UnityEngine.UnitySynchronizationContext:Exec() (at /Users/builduser/buildslave/unity/build/Runtime/Export/Scripting/UnitySynchronizationContext.cs:74)
> UnityEngine.UnitySynchronizationContext:ExecuteTasks() (at /Users/builduser/buildslave/unity/build/Runtime/Export/Scripting/UnitySynchronizationContext.cs:94)
>  
> (Filename: /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Worker/WorkerConnector.cs Line: 324)

However in the same file above there is a line which mach a regex:
> PlayerConnection initialized network socket : 0.0.0.0 55333

**Comment by jamiebrynes7** [comment]
Right, but the exception is caught and logged (what you see in the snippet you posted above), so I'm unsure why (or if) that would cause the connection failure. 

Do you see any other errors in the output logs?

**Comment by v1ctor** [comment]
@jamiebrynes7 Oh, I think you're right, I've confused the cause. 
Here is logs when launcher cannot connect to the deployment:

> [UnityClient] Unable to create player after 1 attempt.
> 'WorkerType': 'UnityClient'
> UnityEngine.DebugLogHandler:Internal_Log(LogType, LogOption, String, Object)
> UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
> UnityEngine.Logger:Log(LogType, Object, Object)
> Improbable.Gdk.Core.ForwardingDispatcher:HandleLog(LogType, LogEvent) (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Logging/ForwardingDispatcher.cs:80)
> Improbable.Gdk.PlayerLifecycle.SendCreatePlayerRequestSystem:RetryCreatePlayerRequest() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.playerlifecycle@0.3.1/Systems/PlayerCreation/SendCreatePlayerRequestSystem.cs:164)
> Improbable.Gdk.PlayerLifecycle.SendCreatePlayerRequestSystem:HandlePlayerCreation() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.playerlifecycle@0.3.1/Systems/PlayerCreation/SendCreatePlayerRequestSystem.cs:206)
> Improbable.Gdk.PlayerLifecycle.SendCreatePlayerRequestSystem:OnUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.playerlifecycle@0.3.1/Systems/PlayerCreation/SendCreatePlayerRequestSystem.cs:224)
> Unity.Entities.ComponentSystem:InternalUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:800)
> Unity.Entities.ComponentSystemBase:Update() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:284)
> Unity.Entities.ComponentSystemGroup:OnUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystemGroup.cs:602)
> Unity.Entities.ComponentSystem:InternalUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:800)
> Unity.Entities.ComponentSystemBase:Update() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:284)
> Unity.Entities.ComponentSystemGroup:OnUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystemGroup.cs:602)
> Unity.Entities.ComponentSystem:InternalUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:800)
> Unity.Entities.ComponentSystemBase:Update() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ComponentSystem.cs:284)
> Unity.Entities.DummyDelegateWrapper:TriggerUpdate() (at /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/com.unity.entities@0.1.1-preview/Unity.Entities/ScriptBehaviourUpdateOrder.cs:144)
> 
> (Filename: /Users/bool/gamedev/examples/gdk-for-unity-fps-starter-project/workers/unity/Library/PackageCache/io.improbable.gdk.core@0.3.1/Logging/ForwardingDispatcher.cs Line: 80)

**Comment by jamiebrynes7** [comment]
Okay, that's more like it 😅. So the client-worker connected successfully, but the `CreatePlayer` request to the UnityGameLogic worker failed for one reason or another. That means that either: 

- There is no UnityGameLogic worker to handle the create player request
- The UnityGameLogic worker failed to create the player for your client

Could you check the deployment logs to check for: 

1. The UnityGameLogic worker connecting successfully to the deployment
2. Any error logs relating to the UnityGameLogic worker failing to create the player.

**Comment by v1ctor** [comment]
TBH I don't see a UnityGameLogic worker anywhere, where I can find it?

**Comment by jamiebrynes7** [comment]
You should see a message in the [Logs](https://docs.improbable.io/reference/14.2/shared/operate/logs) for your deployment similar to:

```
Handling connection request for worker of type UnityGameLogic trying to log in: UnityGameLogic0
```

and then:

```
The worker UnityGameLogic0 connected to SpatialOS successfully.
```

If it connected successfully

**Comment by v1ctor** [comment]
Ok, I found the the number of live workers is 0 for all workers. I'll try to restart deployment now to see if it helps.

It strange I didn't see anything related to workers in the Get Started docs.

**Comment by jamiebrynes7** [comment]
The Get Started docs gloss over it a bit to get you into game ASAP. You can see the worker configuration in the [load balancing section](https://docs.improbable.io/reference/14.2/shared/worker-configuration//load-balancing#load-balancing) of the launch configuration (for example - https://github.com/spatialos/gdk-for-unity-fps-starter-project/blob/master/cloud_launch_large.json#L16).

Provided you have built the workers (especially the `UnityGameLogic@Linux` worker), it should work out of the box. You can sanity check this by looking at the zips in the `${SPATIAL_PROJECT_ROOT}/build/assembly/workers` directory

**Comment by v1ctor** [comment]
I think that I found the problem:

> Could not download Artifact(UnityGameLogic@Linux.zip) for Worker(UnityGameLogic). The Runtime will be unable to start workers of this type.: rpc error: code = NotFound desc = failed to get artifact(UnityGameLogic@Linux.zip) for assembly(beta_may_nuts_261_1126_1142), project(beta_may_nuts_261): failed to get artifact(UnityGameLogic@Linux.zip) for assembly(beta_may_nuts_261_1126_1142), project(beta_may_nuts_261)
> error
> 
> failed to get artifact(UnityGameLogic@Linux.zip) for assembly(beta_may_nuts_261_1126_1142), project(beta_may_nuts_261)

And of course there is no such an artifact in the assembly. However, it's not clear how to cross compile for Linux on macOS, or how to configure deployment to use mac from the docs.

**Comment by v1ctor** [comment]
Wow, I'm an idiot, I've build it for local and not for cloud :)

I believe it will work now.

Thank for your help @jamiebrynes7 !

**Comment by v1ctor** [comment]
I don't know if you'd like to do anything with the port error though. Do you want me to create another issue?

**Comment by jamiebrynes7** [comment]
> I believe it will work now.

Awesome! 🎉 Let me know if you run into any more problems! 

> I don't know if you'd like to do anything with the port error though. Do you want me to create another issue?

We've made an internal bug ticket for this, lets leave this issue open to track it publicly. 😄

**Comment by jamiebrynes7** [comment]
This was fixed in #1232 and released in [`v0.3.2`](https://github.com/spatialos/gdk-for-unity/releases/tag/0.3.2)!
